content/wayland-rs-v0.20.markdown

```diff
@@ -0,0 +1,158 @@
+got from working on [smithay](https://github.com/smithay/smithay), using `wayland-server`.
+
+## Summary of the changes
+
+If I had to describe this refactor in a single sentence, I'd say: "don't be afraid to trade
+small runtime costs for large ergonomic improvements". The previous design of the libraries
+tried to remain as low-overhead possible as possible on the wayland C libs. This caused an
+API that was quite convoluted to try to
+[make safe sharing of mutable state](https://docs.rs/token_store/), and at 
+some points not very different from manually re-implementing a v-table. Other issues more
+specific to smithay resulted in a large accumulation of template parameters on types that need
+to be named and stored. All these issues made the API quite tedious to use overall.
+
+As a result, for this new version I decided to embrace trait objects, `Rc`, and `Arc`.
+Additionaly, the API redesign was made focusing on *the spirit* of the wayland protocol,
+rather than trying to stay close to *the C API*.
+
+To list a few clear gains from this rework:
+
+- The scanner (crate that generate API code from the XML protocol files) almost no longer
+  needs to generate `unsafe { }` blocks in its generated code
+- The long standing issue of handling both `Send` and non-`Send` shared state is finally
+  solved (see next section)
+- The ergonomics and flexibility of specifying callbacks for wayland events has been greatly
+  improved
+- Overall better interaction with objects from the C world, if wayland-rs is used in
+  combination with C libraries
+
+## The new callback specification
+
+The most major API change is on the way callbacks are handled. Previously you had to provide
+a set of freestanding functions that would be given access to some state data for each event,
+with the difficulties of sharing state between an undeterminate set of callbacks a priori...
+
+Now each time a new protocol object is created, you'll receive it as a `NewProxy`/`NewResource`
+object (depending on whether you are client side or server side), and will have to implement
+it before accessing the real `Proxy`/`Resource` and being able to use it. Implementing an
+object consists in providing a type implementing the appropriate `Implementation<Meta, Msg>`
+trait.
+
+This trait has a very simple definition:
+
+```rust
+pub trait Implementation<Meta, Msg> {
+    fn receive(&mut self, msg: Msg, meta: Meta);
+}
+```
+
+Here the `Meta` type parameter represent some metadata associated with the message you
+receive (often a `Proxy`/`Resource` handle to the wayland object receiving the message),
+while `Msg` is a type representing the message itself (often an enum of the possible wayland
+messages this object can receive).
+
+The convenience is that `Implementation<Meta, Msg>` is automatically implemented for all
+type that implement `FnMut(Msg, Meta)`. As such, you can easily provide closures as
+implementations, with all the ergonomics of capturing values and code compacity that
```

Comment:
"compacity" doesn't exist. There is "compactness", but "terseness" would probably be better.